src/mongoc/mongoc-topology.c

```diff
@@ -0,0 +1,664 @@
+   const mongoc_host_list_t *hl;
+
+   bson_return_val_if_fail(uri, NULL);
+
+   topology = bson_malloc0(sizeof *topology);
+
+   /*
+    * Not ideal, but there's no great way to do this.
+    * Base on the URI, we assume:
+    *   - if we've got a replicaSet name, initialize to RS_NO_PRIMARY
+    *   - otherwise, if the seed list has a single host, initialize to SINGLE
+    *   - everything else gets initialized to UNKNOWN
+    */
+   if (mongoc_uri_get_replica_set(uri)) {
+      init_type = MONGOC_TOPOLOGY_RS_NO_PRIMARY;
+   } else {
+      hl = mongoc_uri_get_hosts(uri);
+      if (hl->next) {
+         init_type = MONGOC_TOPOLOGY_UNKNOWN;
+      } else {
+         init_type = MONGOC_TOPOLOGY_SINGLE;
+      }
+   }
+
+   mongoc_topology_description_init(&topology->description, init_type, NULL);
+   topology->description.set_name = bson_strdup(mongoc_uri_get_replica_set(uri));
+
+   topology->scanner = mongoc_topology_scanner_new (_mongoc_topology_scanner_cb,
+                                                    topology);
+   topology->users = 0;
+   topology->uri = uri;
+   topology->single_threaded = true;
+
+   /* TODO make SS timeout configurable on client */
+   topology->timeout_msec = 30000;
+
+   mongoc_mutex_init (&topology->mutex);
+   mongoc_cond_init (&topology->cond_client);
+   mongoc_cond_init (&topology->cond_server);
+
+   for ( hl = mongoc_uri_get_hosts (uri); hl; hl = hl->next) {
+      mongoc_topology_description_add_server (&topology->description,
+                                              hl->host_and_port,
+                                              &id);
+      mongoc_topology_scanner_add (topology->scanner, hl, id);
+   }
+
+   return topology;
+}
+
+/*
+ *-------------------------------------------------------------------------
+ *
+ * mongoc_topology_grab --
+ *
+ *       Increments the ref counter in @topology.
+ *
+ * Returns:
+ *       None.
+ *
+ * Side effects:
+ *       None.
+ *
+ *-------------------------------------------------------------------------
+ */
+void
+mongoc_topology_grab (mongoc_topology_t *topology)
+{
+   bson_return_if_fail(topology);
+
+   mongoc_mutex_lock (&topology->mutex);
+   topology->users++;
+   mongoc_mutex_unlock (&topology->mutex);
+}
+
+/*
+ *-------------------------------------------------------------------------
+ *
+ * mongoc_topology_release --
+ *
+ *       Decrements ref count of @topology. If number falls to < 1,
+ *       destroys this mongoc_topology_t.  Treat this as destroy, and do not
+ *       attempt to use @topology after calling.
+ *
+ * Returns:
+ *       None.
+ *
+ * Side effects:
+ *       May destroy @topology.
+ *
+ *-------------------------------------------------------------------------
+ */
+void
+mongoc_topology_release (mongoc_topology_t *topology)
+{
+   bson_return_if_fail(topology);
+
+   mongoc_mutex_lock (&topology->mutex);
+   topology->users--;
+   if (topology->users < 1) {
+      mongoc_mutex_unlock (&topology->mutex);
```

Comment:
"It shall be safe to destroy an initialized mutex that is unlocked. Attempting to destroy a locked mutex results in undefined behavior. " -- http://linux.die.net/man/3/pthread_mutex_destroy

But I meant calling mongoc_mutex_unlock() as last thing before mongoc_mutex_destroy().

I imagine if something happens between the unlock and the destroy, like @hanumantmk is suggesting
Or, someone in the future accidentally calls mongoc_topology_destroy with a locked mutex...

Its likely fine as-is, was just curious why the lock was released and then lot of work is done and then the mutex destroyed - which upon cursorary thought looked a little dangerous. Given it is about to get destroyed anyway, I figured there is no harm in keeping the lock until its destroyed.